group/filippo_ed25519/point.go

```diff
@@ -0,0 +1,195 @@
+package filippo_ed25519
+
+import (
+	"crypto/cipher"
+	"encoding/hex"
+	"errors"
+	filippo_ed25519 "filippo.io/edwards25519"
+	"go.dedis.ch/kyber/v3"
+	"io"
+)
+
+type Point struct {
+	point *filippo_ed25519.Point
+}
+
+func (p *Point) Equal(s2 kyber.Point) bool {
+	return p.point.Equal(s2.(*Point).point) == 1
+}
+
+func (p *Point) Null() kyber.Point {
+	p.point = filippo_ed25519.NewIdentityPoint()
+	return p
+}
+
+func (p *Point) Base() kyber.Point {
+	p.point = filippo_ed25519.NewGeneratorPoint()
+	return p
+}
+
+func (p *Point) Pick(rand cipher.Stream) kyber.Point {
+	return p.Embed(nil, rand)
+}
+
+func (p *Point) Set(a kyber.Point) kyber.Point {
+	if p.point == nil {
+		p.point = new(filippo_ed25519.Point)
+	}
+	p.point.Set(a.(*Point).point)
+	return p
+}
+
+func (p *Point) Clone() kyber.Point {
+	p2 := new(Point)
+	p2.point = new(filippo_ed25519.Point)
+	p2.point.Set(p.point)
+	return p2
+}
+func (p *Point) Add(a, b kyber.Point) kyber.Point {
+	if p.point == nil {
+		p.point = new(filippo_ed25519.Point)
+	}
+	p.point.Add(a.(*Point).point, b.(*Point).point)
+	return p
+}
+
+func (p *Point) Sub(a, b kyber.Point) kyber.Point {
+	if p.point == nil {
+		p.point = new(filippo_ed25519.Point)
+	}
+	p.point.Subtract(a.(*Point).point, b.(*Point).point)
+	return p
+}
+
+func (p *Point) Neg(a kyber.Point) kyber.Point {
+	if p.point == nil {
+		p.point = new(filippo_ed25519.Point)
+	}
+	p.point.Negate(a.(*Point).point)
+	return p
+}
+
+func (p *Point) Mul(a kyber.Scalar, b kyber.Point) kyber.Point {
+	if p == nil || p.point == nil {
+		p.point = new(filippo_ed25519.Point)
+	}
+	if b == nil || b.(*Point).point == nil {
+		p.point = p.point.ScalarBaseMult(a.(*Scalar).scalar)
+	} else {
+		p.point.ScalarMult(a.(*Scalar).scalar, b.(*Point).point)
+	}
+	return p
+}
+
+func (p *Point) EmbedLen() int {
+	// Reserve the most-significant 8 bits for pseudo-randomness.
+	// Reserve the least-significant 8 bits for embedded data length.
+	// (Hopefully it's unlikely we'll need >=2048-bit curves soon.)
+	return (255 - 8 - 8) / 8
+}
+
+func (p *Point) Embed(data []byte, rand cipher.Stream) kyber.Point {
+
+	// How many bytes to embed?
+	dl := p.EmbedLen()
+	if dl > len(data) {
+		dl = len(data)
+	}
+
+	p.point = new(filippo_ed25519.Point)
+
+	for {
+		// Pick a random point, with optional embedded data
+		var b [32]byte
+		rand.XORKeyStream(b[:], b[:])
+		if data != nil {
+			b[0] = byte(dl)       // Encode length in low 8 bits
+			copy(b[1:1+dl], data) // Copy in data to embed
+		}
+
+		_, err := p.point.SetBytes(b[:])
+		if err != nil {
+			continue
```

Comment:
it seems this would produce an infinite loop with a maliciously crafted `data` input, as:
- `SetBytes` returns an error if `b[:]` does not represent a valid point on the curve
- `b[:]` is crafted by copying bytes from `data` and `data`'s length (or a constant, `p.EmbedLen()`).
- on failure, we repeat the same deterministic operations